### ISSUE TITLE
NullPointerException when using Akka Remote

### ISSUE THREAD

**Issue by martijnblankestijn** [kamon-io/kamon-akka-remote#9]
Hi,

I am trying to use Kamon 1.0.0 with Akka 2.5.8 with clustering/persistence/http/remote.

The project is on https://github.com/martijnblankestijn/akka-avro-serialization-poc/tree/feature/kamon with the 3505bac5eb5e7edccf8e393b13052d9db7da8134 commit with the changes.

When running the server I get the following error:
```
2018-01-24 20:58:01,034 - kamon.prometheus.PrometheusReporter -> INFO [kamon.prometheus.PrometheusReporter] PrometheusReporter - Started the embedded HTTP server on http://0.0.0.0:9095
2018-01-24 20:58:02,320 - kamon.zipkin.ZipkinReporter -> WARN [main] ZipkinReporter - For full Zipkin compatibility enable `kamon.trace.join-remote-parents-with-same-span-id` to preserve span id across client/server sides of a Span.
2018-01-24 20:58:02,335 - kamon.zipkin.ZipkinReporter -> INFO [kamon.zipkin.ZipkinReporter] ZipkinReporter - Started the Zipkin reporter.
2018-01-24 20:58:09,219 - akka.event.slf4j.Slf4jLogger -> INFO [appointmentSystem-akka.actor.default-dispatcher-3] Slf4jLogger - Slf4jLogger started
2018-01-24 20:58:09,239 - akka.event.EventStream -> DEBUG[appointmentSystem-akka.actor.default-dispatcher-3] EventStream - logger log1-Slf4jLogger started
2018-01-24 20:58:09,248 - akka.event.EventStream -> DEBUG[appointmentSystem-akka.actor.default-dispatcher-3] EventStream - Default Loggers started
2018-01-24 20:58:10,599 - akka.remote.Remoting -> INFO [appointmentSystem-akka.actor.default-dispatcher-3] Remoting - Starting remoting
2018-01-24 20:58:10,658 - akka.actor.ActorSystemImpl -> ERROR[appointmentSystem-akka.actor.default-dispatcher-4] ActorSystemImpl - Uncaught error from thread [appointmentSystem-akka.actor.default-dispatcher-3]
java.lang.NullPointerException: null
	at akka.kamon.instrumentation.ActorMonitors$$anon$1.processMessage(ActorMonitor.scala:122)
	at akka.kamon.instrumentation.ActorCellInstrumentation.aroundBehaviourInvoke(ActorInstrumentation.scala:45)
	at akka.actor.ActorCell.invoke(ActorCell.scala:489)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at kamon.executors.Executors$InstrumentedExecutorService$$anon$7.run(Executors.scala:270)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:43)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

The NullPointerException happens on
```
def processMessage(pjp: ProceedingJoinPoint, envelopeContext: TimestampedContext, envelope: Envelope): AnyRef = {
  processedMessagesCounter.increment()

==>  Kamon.withContext(envelopeContext.context) {
    pjp.proceed()
  }
}
```
Looking further 
```
@Around("invokingActorBehaviourAtActorCell(cell, envelope)")
def aroundBehaviourInvoke(pjp: ProceedingJoinPoint, cell: ActorCell, envelope: Envelope): Any = {
  actorInstrumentation(cell).processMessage(pjp, envelope.asInstanceOf[InstrumentedEnvelope].timestampedContext(), envelope)
}
```
the envelope.asInstanceOf[InstrumentedEnvelope].timestampedContext() should deliver null.

I traced the startup sequence with the debugger. A couple of envelopes are succesful processed:

- Envelope (InitializeLogger(akka.event.EventStream@123d655e),Actor[akka://appointmentSystem/temp/$a])
- Envelope (Terminated(Actor[akka://appointmentSystem/user]),Actor[akka://appointmentSystem/user])
- Envelope(Debug(EventStream(akka://appointmentSystem),class akka.event.EventStream,logger log1-Slf4jLogger started),Actor[akka://appointmentSystem/deadLetters])
- Envelope (Internals(akka.remote.Remoting@62f769f8,akka.serialization.Serialization@27000903,Actor[akka://appointmentSystem/remote]),Actor[akka://appointmentSystem/deadLetters])
- Envelope (Info(akka.remote.Remoting,class akka.event.DummyClassForStringSources,Starting remoting),Actor[akka://appointmentSystem/deadLetters])

Sidenote: all the classes within the Envelope are Scala case classes and the attribute timestampedContext within it has a value.

The NPE occurs on the RegisterTerminationHook which is a case object (which might or might not be relevant). In the instrumented version the timestampedContext is null and this probably causes the TimestampedContext to be null.

**Comment by ivantopo** [comment]
Hey @martijnblankestijn, first of all: thanks for such a detailed issue report, this helped *a lot* to get to the issue quickly :)

This particular issue is related to having `akka.actor.serialize-messages=on`, which you do in your project. This makes Akka roundtrip the message and create [a copy](https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/actor/dungeon/Dispatch.scala#L153) of the envelope in the process, thus, loosing the context that was there. In theory that setting should only be on for testing purposes and I tried turning it off and starting your project and it didn't throw the exception anymore, but didn't get into much more testing in there.

I'll put a bit of instrumentation around that method to ensure that if the the context gets copied accordingly.

**Comment by ivantopo** [comment]
Fixed on `kamon-akka` by instrumenting the `.copy` method on the envelope (see https://github.com/kamon-io/kamon-akka/commit/a044c45062347e28c193e37afbe8a318ac430f64). I published `1.0.1-a044c45062347e28c193e37afbe8a318ac430f64` on our snapshots repo if you want to give it a try. Will probably release this over the weekend.

Thanks for reporting the issue and please let us know if you need anything else :slightly_smiling_face:

**Comment by martijnblankestijn** [comment]
Thx @ivantopo, quick reaction!

Tested it and works like a charm! For the serialization I was testing I wanted to force the serialization of all messages even in a single server, that apparently triggered the bug.

**Comment by ivantopo** [comment]
This was released in v1.0.1!
